lib/regexp/regexp.go

```diff
@@ -0,0 +1,290 @@
+// Copyright 2021 The Bazel Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package regexp provides functions related to regular expressions.
+package regexp // import "go.starlark.net/lib/regexp"
+
+import (
+	"fmt"
+	"regexp"
+	"sort"
+	"strings"
+
+	"go.starlark.net/starlark"
+	"go.starlark.net/starlarkstruct"
+)
+
+// Module regexp is a Starlark module of functions related to regular expressions.
+// The module defines the following functions:
+//
+//     compile(pattern) - Compiles a pattern in RE2 syntax (https://github.com/google/re2/wiki/Syntax) to a value of type 'regexp'.
+//                        Each call to compile returns a distinct regexp value.
+//                        A regexp value can be used for matching using its matches, find, find_all and other methods.
+//
+//     regexp.find(src) - Returns a string holding the text of the leftmost match in the given string of the regular expression regexp.
+//                         The result is "" if there is no match or if the pattern successfully matches an empty string.
+//
+//     regexp.find_all(src, max) - Returns a new, mutable list of all successive matches of the regular expression regexp. An empty list indicates
+//                                 no match. If max > 0, at most max strings are returned. If max == 0,
+//                                 an empty list is returned. If max < 0, all strings are returned.
+//                                 The parameter max is optional: by default no limit is applied.
+//
+//     regexp.find_submatches(src) - Returns a new, mutable list of strings holding the text of the leftmost match of the regular expression regexp
+//                                   in the given string and the matches, if any, of its subexpressions. An empty list indicates
+//                                   no match.
+//
+//     regexp.matches(src) - Reports whether the given string contains any match of the regular expression regexp.
+//
+//     regexp.replace_all(src, repl) - Returns a copy of the given string, replacing matches of the regular expression regexp
+//                                     with the replacement string repl. Inside repl, backslash-escaped digits (\1 to \9) can be
```

Comment:
Which regexp package's design is this feature following? Python?  I can't tell whether convertReplacementFunction is correct without a more detailed description of its intention.

I wonder whether the feature is needed.

Reply:
This comes from the C++ implementation of re2 https://github.com/google/re2/blob/master/re2/re2.h#L444-L459

Reply:
What `convertReplacementPattern` does is basically converting a replacement pattern composed of backreferences of type `\i` (where i is the index of the group) into a replacement pattern composed of backreferences of type `${i}`.

Backreferences of type backslash-escaped digits are what is we have in python as you can see in [this example ](https://regex101.com/r/jprisL/1)

Reply:
You have the same logic in the python lib https://docs.python.org/3/library/re.html#re.sub

Reply:
What are the merits of using `\i` syntax instead of Go's `$i` or `${i}`? I understand staying closer to what a Python programmer may be familiar with, but then we're already using Go as the inspiration for the method names and signatures.

Can `\i` notation support more than 9 subexpressions?

Make sure to clarify how a literal `\` can be represented. Consider including `replace_all_literal()`, though that can already be achieved by using `replace_all(src, lambda x: "literal_result")`.

FWIW, `re.sub` has a (to me) ridiculous behavior of expanding substring values representing backslash escapes into corresponding values. I.e., `re.sub` turns `"\\n"` in the replacement text into `"\n"` when substituted. So I'm not in favor of mirroring Python's re API just for mere parity.